docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+                - name: dataset-initializer
+                  image: docker.io/kubeflow/dataset-initializer
+                  env:
+                    - name: DATASET_PROVIDER
+                      value: hf
+                    - name: REPO_ID
+                      value: tatsu-lab/alpaca
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+                - name: model-initializer
+                  image: docker.io/kubeflow/model-initializer
+                  env:
+                    - name: MODEL_PROVIDER
+                      value: hf
+                    - name: REPO_ID
+                      value: google/gemma-7b
+                    - name: TRANSFORMER_TYPE
+                      value: AutoModelForCausalLM
+                  volumeMounts:
+                    - mountPath: /workspace/model
+                      name: model-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+    - name: Node
+      template:
+        spec:
+          template:
+            spec:
+              containers:
+                - name: trainer
+                  image: docker.io/kubeflow/llm-trainer
+                  env:
+                    - name: MASTER_ADDR
+                      value: "pytorch-node-0-0.pytorch"
+                    - name: MASTER_PORT
+                      value: 29400
+                    - name: TRANSFORMER_TYPE
+                      value: AutoModelForCausalLM
+                    - name: LORA_CONFIG
+                      value: |
+                        {"peft_type": "LORA", "r": 8, "lora_alpha": 16}
+                  command:
+                    - torchrun hf_llm_training.py
+                  resources:
+                    limits:
+                      nvidia.com/gpu: 2
+                  volumeMounts:
+                    - mountPath: /workspace/dataset
+                      name: dataset-initializer
+                    - mountPath: /workspace/model
+                      name: model-initializer
+              volumes:
+                - name: dataset-initializer
+                  persistentVolumeClaim:
+                    claimName: dataset-initializer
+                - name: model-initializer
+                  persistentVolumeClaim:
+                    claimName: model-initializer
+```
+
+#### MPI Runtime
+
+For MPI, we can add support the `DeepSpeed` runtimes.
+
+Example of simple OpenMPI runtime:
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: ClusterTrainingRuntime
+metadata:
+  name: mpi-simple
+spec:
+  mlSpec:
+    mpi:
+      mpiImplementation: OpenMPI
+      numProcPerNode: 5
+  numNodes: 5
+  replicatedJobs:
+    - name: Launcher
```

Comment:
I'm a little worried about this design because MPI doesn't always need a node that is explicitly for a launcher - that's a design that the MPI operator has taken (and makes sense, it's part of how MPI design is described), but for example, the lead broker of the flux operator does the launch with that node handling the same work as a worker.

Reply:
Follow up question - could the Flux Operator be handled here? High level, it's an MPI launcher (but actually is an entire HPC cluster that simply can be used to run one MPI job - it's an indexed job + headless service under the hood, with flux orchestration and nodes connected via a tree based overlay network with zeromq).

Reply:
>I'm a little worried about this design because MPI doesn't always need a node that is explicitly for a launchel

Yes, I know we have an API to run launcher as worker in MPI-Operator: https://github.com/kubeflow/mpi-operator/blob/master/pkg/apis/kubeflow/v2beta1/types.go#L161.

I want to hear @tenzen-y and @alculquicondor opinion on this. We didn't get a chance to deeply discuss how MPI Runtime should look like.

Reply:
>Follow up question - could the Flux Operator be handled here? High level, it's an MPI launcher (but actually is an entire HPC cluster that simply can be used to run one MPI job - it's an indexed job + headless service under the hood, with flux orchestration and nodes connected via a tree based overlay network with zeromq)

Yes, I think that is possible. Do you know if we can construct the Flux cluster for batch using JobSet spec or we need to use Flux operator mini-cluster spec ?
```
 JobSetSpec *batchv1.JobSetSpec `json:",inline"`
 ```

Reply:
The Flux Operator (under the hood) is an indexed job, so yes I could very easily turn that into a JobSet. I actually did a design with JobSet early on, but didn't wind up converting entirely for simple reasons - JobSet wasn't part of Kubernetes core (and would require the user to install an additional thing), and JobSet makes the DNS names much longer, and (in my testing) it often led to errors depending on the name of the job that I chose. Since I just needed the IndexedJob and then the headless service, it was more straight forward to just implement that.

Please put me down to implement Flux Framework into this setup - the nested hierarchical scheduler can handle a lot of really cool stuff within a single job, and the zeromq bootstrap is an improvement over ssh.

Reply:
Would be interesting to see the benchmarks to compare zmq over ssh for distributed model training. cc @tenzen-y

Reply:
I can definitely help with that, if there is interest to run experiments. We compared an HPC proxy app (LAMMPS) to the MPI operator, now that was a long time ago (early 2023) and there was about a 5% difference in means, and you can imagine that would compound with runtime (our runs were very short). That said, both have changed immensely since then!

I have a hard time getting the MPI Operator working, but if we wanted to decide on a workflow, I could offer to set it up and run in the Flux Operator. We would need to choose a cloud that has a suitable network and resources for the work. If it's ML then probably Google would work - I have credits currently.

Reply:
I still think that both MPIJob modes (runLauncherAsWorker and specific launcher) would be helpful for the GPU and CPU mixed clusters. Additionally, the launcher allows us to easily migrate from MPIJob v2 to TrainJob.